HDDS-12175. Audit logs in SCM shouldn't print delete txns

Tejaskriya opened the pull request:
## What changes were proposed in this pull request?

As a part of SCM audits for datanode's heartbeats, the command.toString for the commands being sent to the datanodes are logged. For most commands this results in a short string, but for deleteBlocks command, all the delete transactions are also added. This leads to a very verbose audit log. 
SCM should add only the metadata response to the logs appropriately for each of the commands. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-12175

## How was this patch tested?

Tested manually.
Before the change:
```
2025-02-05 17:12:56,748 | INFO  | SCMAudit | user=hadoop | ip=172.18.0.10 | op=SEND_HEARTBEAT {datanodeUUID=b8ca7816-1298-4f9b-b962-a8b43351fb30, term=3, command=[commandType: deleteBlocksCommand deleteBlocksCommandProto { deletedBlocksTransactions { txID: 4 containerID: 4 localID: 115816896921600004 localID: 115816896921600005 localID: 115816896921600006 count: 0 } cmdId: 1738775521193 } term: 3 encodedToken: "" deadlineMsSinceEpoch: 0 ]} | ret=SUCCESS |
```
After the change:
```
2025-02-12 07:33:13,336 | INFO  | SCMAudit | user=hadoop | ip=172.18.0.5 | op=SEND_HEARTBEAT {datanodeUUID=6ad9c70a-b177-4e4a-b562-02019e68b957, term=4, command=[commandType: deleteBlocksCommand deleteTransactionsCount: 1 cmdID: 1739345533733 encodedToken: "" deadlineMsSinceEpoch: 0]} | ret=SUCCESS |
```
The debug log added:
```
2025-02-12 07:33:13,335 [IPC Server handler 10 on default port 9861] DEBUG server.SCMDatanodeHeartbeatDispatcher: Heartbeat dispatched: datanode=6ad9c70a-b177-4e4a-b562-02019e68b957, Commands= [deleteBlocksCommand: cmdID: 1739345533733, encodedToken: "", term: 4, deadlineMsSinceEpoch: 0, deletedBlocksTransaction: [ txnID:3,containerID:3,deleteBlockCount:3,count:0]]
```

adoroszlai commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeProtocolServer.java:
```diff
@@ -280,6 +280,22 @@ public static SCMRegisteredResponseProto getRegisteredResponse(
+    for (SCMCommandProto cmd : cmds) {
+      if (cmd.getCommandType().equals(deleteBlocksCommand)) {
+        auditMap.append("commandType:" + cmd.getCommandType());
+        auditMap.append(" No. of deleteTransactions:");
```
Avoid space in key:

```suggestion
        auditMap.append(" deleteTransactionsCount:");
```

adoroszlai commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeProtocolServer.java:
```diff
@@ -280,6 +280,22 @@ public static SCMRegisteredResponseProto getRegisteredResponse(
+    auditMap.append('[');
+    for (SCMCommandProto cmd : cmds) {
+      if (cmd.getCommandType().equals(deleteBlocksCommand)) {
+        auditMap.append("commandType:" + cmd.getCommandType());
```
Append separately:

```suggestion
        auditMap.append("commandType:").append(cmd.getCommandType());
```

adoroszlai commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeProtocolServer.java:
```diff
@@ -280,6 +280,22 @@ public static SCMRegisteredResponseProto getRegisteredResponse(
+      if (cmd.getCommandType().equals(deleteBlocksCommand)) {
+        auditMap.append("commandType:" + cmd.getCommandType());
+        auditMap.append(" No. of deleteTransactions:");
+        auditMap.append(cmd.getDeleteBlocksCommandProto().getDeletedBlocksTransactionsList().size());
+      } else {
```
I guess some separator is also needed between commands, maybe between command attributes, too.  The goal is to most closely match previous output while ommitting unwanted details.

adoroszlai reviewed (commented):
Thanks @Tejaskriya for the patch.

Please provide examples of audit log before/after the change, can take it from acceptance test logs ([before](https://github.com/apache/ozone/actions/runs/13155119418#artifacts), [after](https://github.com/Tejaskriya/ozone/actions/runs/13131442815#artifacts)).

Tejaskriya commented:
@adoroszlai Thanks for the review! I have made the changes and added the old and new audit log in the PR description.

adoroszlai commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeProtocolServer.java:
```diff
@@ -280,6 +280,29 @@ public static SCMRegisteredResponseProto getRegisteredResponse(
+      } else {
+        auditMap.append(cmd);
+      }
+      auditMap.append(" , ");
```
```suggestion
      auditMap.append(", ");
```

adoroszlai commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeProtocolServer.java:
```diff
@@ -280,6 +280,29 @@ public static SCMRegisteredResponseProto getRegisteredResponse(
+      }
+      auditMap.append(" , ");
+    }
+    int len = auditMap.toString().length();
```
Let's not call `toString()` twice.

```suggestion
    int len = auditMap.length();
```

adoroszlai commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeProtocolServer.java:
```diff
@@ -280,6 +280,29 @@ public static SCMRegisteredResponseProto getRegisteredResponse(
+      auditMap.append(" , ");
+    }
+    int len = auditMap.toString().length();
+    if (len > 3) {
+      auditMap.deleteCharAt(len - 1);
+      auditMap.deleteCharAt(len - 2);
+      auditMap.deleteCharAt(len - 3);
+    }
```
I think this can be simplified.

(3 should be 2 if ` ,` is changed to `, `.)

```suggestion
    if (len > 3) {
      auditMap.delete(len - 3, len);
    }
```

adoroszlai reviewed (commented):
Thanks for updating the patch.

Tejaskriya commented:
@adoroszlai I have addressed all the comments, Thanks for suggesting these improvements!

nandakumar131 reviewed (commented):
Thanks @Tejaskriya for working on this.
Overall the change looks good. 
Let's retain `cmdId`, `encodedToken` and `deadlineMsSinceEpoch` in the audit log for `deleteBlocksCommand` as well.

Also, please add a debug level log which prints the `transactions`. We don't have to print the block IDs (Local IDs), instead we can print the total number of blocks in a container.

nandakumar131 commented:
You can add the debug log in `SCMDatanodeHeartbeatDispatcher#dispatch`.

Tejaskriya commented:
Thanks for the review @nandakumar131 , I have added the extra details to the audit log and also added a debug log in `SCMDatanodeHeartbeatDispatcher#dispatch`.
The updated output is added to the PR description

nandakumar131 commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeHeartbeatDispatcher.java:
```diff
@@ -200,6 +203,34 @@ public List<SCMCommand> dispatch(SCMHeartbeatRequestProto heartbeat) {
+    if (LOG.isDebugEnabled()) {
+      StringBuilder allCommands = new StringBuilder();
+      for (SCMCommand cmd : commands) {
+        allCommands.append("cmdID: ").append(cmd.getId());
+        allCommands.append(" encodedToken: \"").append(cmd.getEncodedToken()).append("\"");
+        allCommands.append(" term: ").append(cmd.getTerm());
+        allCommands.append(" deadlineMsSinceEpoch: ").append(cmd.getDeadline());
+        if (cmd.getType().equals(deleteBlocksCommand)) {
+          DeleteBlocksCommand delCmd = (DeleteBlocksCommand) cmd;
+          allCommands.append(" deleteBlocksTransactions: {");
+          for (DeletedBlocksTransaction txn : delCmd.blocksTobeDeleted()) {
+            allCommands.append(" txnId:").append(txn.getTxID());
+            allCommands.append(" containerID:").append(txn.getContainerID());
+            allCommands.append(" deleteBlockCount:").append(txn.getLocalIDCount());
+            allCommands.append(" count:").append(txn.getCount());
+            allCommands.append(",");
+          }
+          allCommands.deleteCharAt(allCommands.length() - 1);
+          allCommands.append(" }");
+        } else {
+          allCommands.append(" ").append(cmd.getProto());
+        }
+        allCommands.append(", ");
+      }
+      int len = allCommands.length();
+      allCommands.delete(len - 2, len);
```
`SCMCommand` is not a protobuf object, you can override `toString` method in all the `SCMCommand` implementation and just logging the `SCMCommand#toString` here should be enough.

Tejaskriya replied:
I have changed the toString method for all commands to print the required information along with the common details like cmdID, deadline, term and encodedToken in the latest commit.

nandakumar131 commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMDatanodeHeartbeatDispatcher.java:
```diff
@@ -200,6 +200,17 @@ public List<SCMCommand> dispatch(SCMHeartbeatRequestProto heartbeat) {
       }
     }
+    if (LOG.isDebugEnabled()) {
+      StringBuilder allCommands = new StringBuilder();
+      for (SCMCommand cmd : commands) {
+        allCommands.append(cmd).append(", ");
+      }
+      int len = allCommands.length();
+      if (len > 2) {
+        allCommands.delete(len - 2, len);
+      }
+      LOG.debug("Heartbeat dispatched: datanode=" + datanodeDetails.getUuid() + ", Commands= [" + allCommands + "]");
```
```suggestion
      LOG.debug("Heartbeat dispatched: datanode=" + datanodeDetails.getUuid() + ", Commands= " + commands);
```

nandakumar131 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/protocol/commands/DeleteBlocksCommand.java:
```diff
@@ -68,4 +68,25 @@ public DeleteBlocksCommandProto getProto() {
+        .append(", term: ").append(getTerm())
+        .append(", deadlineMsSinceEpoch: ").append(getDeadline())
+        .append(", deletedBlocksTransaction: [");
+    for (DeletedBlocksTransaction txn : blocksTobeDeleted) {
+      sb.append(" txnID:").append(txn.getTxID())
+          .append(",containerID:").append(txn.getContainerID())
+          .append(",deleteBlockCount:").append(txn.getLocalIDCount())
+          .append(",count:").append(txn.getCount())
+          .append(",");
+    }
+    sb.deleteCharAt(sb.length() - 1);
```
```suggestion
    for (DeletedBlocksTransaction txn : blocksTobeDeleted) {
      sb.append("{ txnID: ").append(txn.getTxID())
          .append(", containerID: ").append(txn.getContainerID())
          .append(", deleteBlockCount: ").append(txn.getLocalIDCount())
          .append(", count: ").append(txn.getCount())
          .append("}, ");
    }
    if (!blocksTobeDeleted.isEmpty()) {
      sb.delete(sb.length() - 2, sb.length());
    }
```

Tejaskriya commented:
Thanks for the review @nandakumar131 , I have incorporated the suggestions

nandakumar131 reviewed (approved):
+1, pending CI.

nandakumar131 commented:
Thanks @Tejaskriya for the contribution! Thanks @adoroszlai for the review.
